File: frontend/src/api/kinetics.ts
const API_BASE = '/api/v1/kinetics';


export interface NutritionMeal {
  id: string;
  user_id: string;
  meal_date: string;
  time_str: string;
  name: string;
  weight_g: number;
  calories: number;
  protein: number;
  fat: number;
  carbs: number;
  ingredients: any[];
  created_at: string;
}

export interface DailyNutritionSummary {
  date: string;
  total_calories: number;
  total_protein: number;
  total_fat: number;
  total_carbs: number;
  meals: NutritionMeal[];
}

export interface SegmentData {
  muscle_kg: number;
  muscle_pct: number;
  fat_kg: number;
  fat_pct: number;
}

export interface BiometricsCreate {
  weight_kg?: number;
  sleep_hours?: number;
  fatigue_score?: number;
  notes?: string;
}
export interface RecoveryDataPoint {
  date: string;
  tonnage_kg: number;
  calories: number;
  protein_g: number;
  sleep_hours: number;
  fatigue_score: number;
  recovery_score: number;
}
export interface RecoveryAnalyticsResponse {
  timeline: RecoveryDataPoint[];
  average_sleep: number;
  average_calories: number;
  total_tonnage: number;
  current_recovery_score: number;
}
export interface BiometricsLog {
  id: string;
  weight?: number;
  weight_kg?: number;
  body_fat_percentage?: number;
  fat_mass_kg?: number;
  skeletal_muscle_kg?: number;
  muscle_kg?: number;
  water_l?: number;
  protein_kg?: number;
  minerals_kg?: number;
  visceral_fat_level?: number;
  bmr_kcal?: number;
  bmr?: number;
  tdee?: number;
  bmi?: number;
  
  calories_in?: number;
  protein_g?: number;
  sleep_hours?: number;
  fatigue_score?: number;
  segment_data?: Record<string, SegmentData>;
  
  timestamp: string;
}

export interface ScannedMealData {
  name: string;
  portion_weight_g: number;
  weight_g?: number;
  calories: number;
  protein: number;
  fat: number;
  carbs: number;
  ingredients_detected: string[];
  confidence_note: string;
}

export interface WorkoutGenerateParams {
  location: 'Дом' | 'Зал';
  split_day: string;
  target_split?: string;
}

export interface WorkoutPlanSettingsPayload {
  plan_id?: string;
  location?: string;
  target_split?: string;
  split_day?: string;
  restrictions?: string;
  goals?: string;
}

export interface AthleteProfileData {
  id?: string;
  age: number;
  gender: string;
  height_cm: number;
  target_fat_pct: number;
  target_weight_kg: number;
  goals: string[];
  lagging_muscles: string[];
  training_experience: string;
  last_break: string;
  workout_frequency: number;
  duration_min: number;
  preferred_time: string;
  schedule_days: string[];
  restrictions: string;
  strength_bench: number;
  strength_squat: number;
  strength_deadlift: number;
  pullups_reps: number;
  mobility_squat: number;
  mobility_shoulder: number;
  mobility_bend: number;
}


export interface OneRMDataPoint {
  date: string;
  one_rm_kg: number;
}
export interface ExerciseOneRM {
  exercise_name: string;
  history: OneRMDataPoint[];
}
export interface TonnageDataPoint {
  week_start: string;
  muscle_group: string;
  tonnage_kg: number;
}
export interface OverloadAnalyticsResponse {
  one_rm_top_exercises: ExerciseOneRM[];
  weekly_tonnage: TonnageDataPoint[];
}
export interface WorkoutSet {
  id: string;
  exercise_id: string;
  set_number: number;
  set_type: 'W' | 'N' | 'D' | 'F';
  weight_kg: number;
  reps: number;
  rpe?: number;
  is_completed: boolean;
  previous_weight_kg?: number;
  previous_reps?: number;
}

export interface WorkoutExercise {
  id: string;
  exercise_name: string;
  exercise_type: 'bodyweight' | 'isometric' | 'cardio' | 'intervals';
  sets: number;
  reps_or_duration: string;
  target_muscle_groups: string[];
  rpe_target: number;
  is_completed: boolean;
  order_index: number;
  workout_sets?: WorkoutSet[];
}

export interface MesocycleSettingsPayload {
  start_date: string; // YYYY-MM-DD
  days_of_week: number[];
  target_split: string;
  location: string;
  include_deload: boolean;
  weeks_count: number;
}

export interface WorkoutPlan {
  id: string;
  target_split: string;
  split_type?: string;
  location?: string;
  ai_rationale: string;
  status: 'planned' | 'in_progress' | 'completed';
  exercises: WorkoutExercise[];
  created_at: string;
  scheduled_date?: string;
}

async function request<T>(url: string, options?: RequestInit): Promise<T> {
  const response = await fetch(url, {
    headers: {
      'Content-Type': 'application/json',
      ...options?.headers,
    },
    ...options,
  });

  if (!response.ok) {
    const errorText = await response.text();
    throw new Error(`Ошибка запроса (${response.status}): ${errorText}`);
  }

  return response.json();
}

export interface ExerciseAlternativeItem {
  exercise_name: string;
  exercise_type: string;
  sets: number;
  reps_or_duration: string;
  rpe_target: number;
  target_muscle_groups: string[];
  biomechanical_rationale: string;
}

export const kineticsApi = {
  getProfile: async (): Promise<any> => {
    const res = await fetch('/api/v1/kinetics/profile');
    if (!res.ok) throw new Error('Ошибка загрузки профиля');
    return res.json();
  },

  updateProfile: async (data: Partial<AthleteProfileData>): Promise<AthleteProfileData> => {
    const res = await fetch(`${API_BASE}/profile`, {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data),
    });
    return res.json();
  },

  getBiometrics: async (limit: number = 7): Promise<BiometricsLog[]> => {
    return request<BiometricsLog[]>(`${API_BASE}/biometrics?limit=${limit}`);
  },

  logBiometrics: async (payload: Omit<BiometricsLog, 'id' | 'timestamp'>): Promise<BiometricsLog> => {
    return request<BiometricsLog>(`${API_BASE}/biometrics`, {
      method: 'POST',
      body: JSON.stringify(payload),
    });
  },

  getLatestWorkout: async (): Promise<WorkoutPlan | null> => {
    return request<WorkoutPlan | null>(`${API_BASE}/workouts/latest`);
  },

  generateWorkout: async (params: WorkoutGenerateParams): Promise<WorkoutPlan> => {
    return request<WorkoutPlan>(`${API_BASE}/workouts/generate`, {
      method: 'POST',
      body: JSON.stringify(params),
    });
  },

  toggleExercise: async (exerciseId: string, isCompleted: boolean): Promise<void> => {
    await request<void>(`${API_BASE}/exercises/${exerciseId}`, {
      method: 'PATCH',
      body: JSON.stringify({ is_completed: isCompleted }),
    });
  },

  patchExercise: async (id: string, data: Partial<WorkoutExercise>): Promise<WorkoutExercise> => {
    const res = await fetch(`${API_BASE}/exercises/${id}`, {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data)
    });
    return res.json();
  },

  deleteExercise: async (id: string): Promise<void> => {
    await fetch(`${API_BASE}/exercises/${id}`, { method: 'DELETE' });
  },

  addExercise: async (planId: string, name: string): Promise<WorkoutExercise> => {
    const res = await fetch(`${API_BASE}/exercises`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        plan_id: planId,
        exercise_name: name,
        exercise_type: 'hypertrophy',
        sets: 3,
        reps_or_duration: '10-12',
        rpe_target: 7,
        target_muscle_groups: ['дополнительно']
      })
    });
    return res.json();
  },

  askCoach: async (message: string, planId: string): Promise<WorkoutPlan> => {
    const res = await fetch(`${API_BASE}/workouts/chat`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ message, plan_id: planId })
    });
    return res.json();
  },

  chatWithCoach: async (message: string, planContext?: any): Promise<any> => {
    const res = await fetch('/api/v1/kinetics/workouts/chat', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ message, plan_context: planContext })
    });
    if (!res.ok) throw new Error('Ошибка генерации ответа тренера');
    return res.json();
  },

  getExerciseAlternatives: async (
    exerciseId: string,
    reason: 'joint_pain' | 'no_axial_load' | 'equipment_busy' | 'too_intense' = 'joint_pain'
  ): Promise<ExerciseAlternativeItem[]> => {
    const res = await fetch(`/api/v1/kinetics/exercises/${exerciseId}/alternatives`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ reason })
    });
    if (!res.ok) throw new Error('Ошибка получения альтернатив');
    return res.json();
  },

  swapExercise: async (
    exerciseId: string,
    payload: {
      exercise_name: string;
      exercise_type: string;
      sets: number;
      reps_or_duration: string;
      rpe_target: number;
      target_muscle_groups: string[];
    }
  ): Promise<WorkoutExercise> => {
    const res = await fetch(`/api/v1/kinetics/exercises/${exerciseId}/swap`, {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(payload)
    });
    if (!res.ok) throw new Error('Ошибка замены упражнения');
    return res.json();
  },

  getChatHistory: async (module: 'coach' | 'nutrition'): Promise<any[]> => {
    const res = await fetch(`/api/v1/kinetics/chat/${module}`);
    if (!res.ok) throw new Error('Ошибка загрузки истории чата');
    return res.json();
  },

  summarizeChat: async (module: 'coach' | 'nutrition'): Promise<any> => {
    const res = await fetch(`/api/v1/kinetics/chat/${module}/summarize`, {
      method: 'POST'
    });
    if (!res.ok) throw new Error('Ошибка суммаризации чата');
    return res.json();
  },

  
  getDailyNutrition: async (date: string): Promise<DailyNutritionSummary> => {
    return request<DailyNutritionSummary>(`${API_BASE}/nutrition/daily?date=${date}`);
  },

  addNutritionMeal: async (payload: Omit<NutritionMeal, 'id' | 'user_id' | 'created_at'>): Promise<NutritionMeal> => {
    return request<NutritionMeal>(`${API_BASE}/nutrition/meals`, {
      method: 'POST',
      body: JSON.stringify(payload)
    });
  },

  deleteNutritionMeal: async (mealId: string): Promise<void> => {
    return request<void>(`${API_BASE}/nutrition/meals/${mealId}`, {
      method: 'DELETE'
    });
  },

  scanMealPhoto: async (file: File): Promise<ScannedMealData> => {
    const formData = new FormData();
    formData.append('file', file);
    const res = await fetch(`${API_BASE}/nutrition/scan-photo`, {
      method: 'POST',
      body: formData,
    });
    if (!res.ok) throw new Error('Ошибка сканирования фото');
    return res.json();
  },

  sendNutritionChatMessage: async (payload: { message: string, meals: any[], target_calories: number, current_weight: number }): Promise<{ response: string }> => {
    const res = await fetch(`${API_BASE}/nutrition/chat`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(payload)
    });
    if (!res.ok) throw new Error('Ошибка общения с AI-диетологом');
    return res.json();
  },



  getAnalyticsCorrelation: async (weeks: number = 8): Promise<any[]> => {
    const res = await fetch(`${API_BASE}/analytics/correlation?weeks=${weeks}`);
    if (!res.ok) throw new Error('Ошибка загрузки корреляционной аналитики');
    return res.json();
  },

  getCalendarMonth: async (): Promise<WorkoutPlan[]> => {
    const res = await fetch(`${API_BASE}/calendar/month`);
    if (!res.ok) throw new Error('Ошибка загрузки календаря');
    return res.json();
  },

  updateWorkoutPlanSettings: async (payload: WorkoutPlanSettingsPayload): Promise<WorkoutPlan> => {
    const res = await fetch(`${API_BASE}/workouts/plan`, {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(payload)
    });
    if (!res.ok) throw new Error('Ошибка обновления настроек плана');
    return res.json();
  },

  completeWorkout: async (planId: string): Promise<WorkoutPlan> => {
    const res = await fetch(`${API_BASE}/workouts/complete`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ plan_id: planId })
    });
    if (!res.ok) throw new Error('Ошибка завершения тренировки');
    return res.json();
  },

  rescheduleCalendar: async (days: any[]): Promise<any> => {
    const res = await fetch(`${API_BASE}/calendar/reschedule`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ days })
    });
    if (!res.ok) throw new Error('Ошибка перепланирования календаря');
    return res.json();
  },

  generateMesocycle: async (payload: MesocycleSettingsPayload): Promise<{status: string, message: string}> => {
    const res = await fetch(`${API_BASE}/workouts/generate-mesocycle`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(payload)
    });
    if (!res.ok) throw new Error('Ошибка синтеза мезоцикла');
    return res.json();
  },
      getRecoveryAnalytics: async (): Promise<RecoveryAnalyticsResponse> => {
    const res = await fetch(`${API_BASE}/analytics/recovery`);
    if (!res.ok) throw new Error('Failed to fetch recovery analytics');
    return res.json();
  },
  saveBiometrics: async (data: BiometricsCreate): Promise<void> => {
    const res = await fetch(`${API_BASE}/biometrics`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data),
    });
    if (!res.ok) throw new Error('Failed to save biometrics');
  },
  getOverloadAnalytics: async (): Promise<OverloadAnalyticsResponse> => {
    const res = await fetch(`${API_BASE}/analytics/overload`);
    if (!res.ok) throw new Error('Failed to fetch overload analytics');
    return res.json();
  },
  duplicateWorkoutPlan: async (planId: string, targetDate: string, applyOverload: boolean = false, overloadIncrementKg: number = 1.25): Promise<WorkoutPlan> => {
    const res = await fetch(`${API_BASE}/workouts/${planId}/duplicate`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ target_date: targetDate, apply_overload: applyOverload, overload_increment_kg: overloadIncrementKg })
    });
    if (!res.ok) throw new Error('Failed to duplicate workout plan');
    return res.json();
  },
  createWorkoutSet: async (exerciseId: string, data: Partial<WorkoutSet>): Promise<WorkoutSet> => {
    const response = await fetch(`${API_BASE}/exercises/${exerciseId}/sets`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data)
    });
    if (!response.ok) throw new Error('Failed to create set');
    return response.json();
  },
  
  updateWorkoutSet: async (setId: string, data: Partial<WorkoutSet>): Promise<WorkoutSet> => {
    const response = await fetch(`${API_BASE}/workout-sets/${setId}`, {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data)
    });
    if (!response.ok) throw new Error('Failed to update set');
    return response.json();
  },
  
  deleteWorkoutSet: async (setId: string): Promise<void> => {
    await fetch(`${API_BASE}/workout-sets/${setId}`, { method: 'DELETE' });
  }
};

File: backend/app/schemas/kinetics.py
from pydantic import BaseModel, Field, ConfigDict, AliasChoices
from typing import List, Optional, Dict, Any
from datetime import datetime, date
import uuid
from enum import Enum


class SegmentData(BaseModel):
    muscle_kg: Optional[float] = None
    muscle_pct: Optional[float] = None
    fat_kg: Optional[float] = None
    fat_pct: Optional[float] = None


class BiometricsLogBase(BaseModel):
    weight: Optional[float] = Field(None, validation_alias=AliasChoices("weight", "weight_kg"))
    weight_kg: Optional[float] = Field(None, validation_alias=AliasChoices("weight_kg", "weight"))
    body_fat_percentage: Optional[float] = Field(None, validation_alias=AliasChoices("body_fat_percentage", "body_fat_pct"))
    body_fat_pct: Optional[float] = Field(None, validation_alias=AliasChoices("body_fat_pct", "body_fat_percentage"))
    fat_mass_kg: Optional[float] = None
    skeletal_muscle_kg: Optional[float] = Field(None, validation_alias=AliasChoices("skeletal_muscle_kg", "muscle_kg"))
    muscle_kg: Optional[float] = Field(None, validation_alias=AliasChoices("muscle_kg", "skeletal_muscle_kg"))
    water_l: Optional[float] = None
    protein_kg: Optional[float] = None
    minerals_kg: Optional[float] = None
    visceral_fat_level: Optional[int] = Field(None, validation_alias=AliasChoices("visceral_fat_level", "visceral_fat"))
    visceral_fat: Optional[int] = Field(None, validation_alias=AliasChoices("visceral_fat", "visceral_fat_level"))
    bmr_kcal: Optional[int] = Field(None, validation_alias=AliasChoices("bmr_kcal", "bmr"))
    bmr: Optional[int] = Field(None, validation_alias=AliasChoices("bmr", "bmr_kcal"))
    bmi: Optional[float] = None

    calories_in: Optional[int] = 0
    tdee: Optional[int] = 0
    protein_g: Optional[int] = None
    sleep_hours: Optional[float] = 8.0
    fatigue_score: Optional[int] = 0
    segment_data: Optional[Dict[str, Any]] = Field(default_factory=dict)
    segment_fat_pct: Optional[Dict[str, Any]] = Field(default_factory=dict)
    notes: Optional[str] = None


class BiometricsLogCreate(BiometricsLogBase):
    pass


class BiometricsLogResponse(BiometricsLogBase):
    id: uuid.UUID
    user_id: Optional[uuid.UUID] = None
    timestamp: Optional[datetime] = Field(None, validation_alias=AliasChoices("timestamp", "created_at"))
    created_at: Optional[datetime] = Field(None, validation_alias=AliasChoices("created_at", "timestamp"))

    model_config = ConfigDict(from_attributes=True, populate_by_name=True, extra="ignore")


class WorkoutExerciseBase(BaseModel):
    exercise_name: str
    exercise_type: str = "bodyweight"
    sets: int = 3
    reps_or_duration: str = "10-12"
    rpe: Optional[int] = Field(None, validation_alias=AliasChoices("rpe", "rpe_target"))
    rpe_target: Optional[int] = Field(None, validation_alias=AliasChoices("rpe_target", "rpe"))
    target_muscle_groups: List[str] = Field(default_factory=list)
    is_completed: bool = False
    order_index: int = 0


class WorkoutExerciseCreate(WorkoutExerciseBase):
    pass


class WorkoutExerciseResponse(WorkoutExerciseBase):
    id: uuid.UUID
    workout_id: Optional[uuid.UUID] = Field(None, validation_alias=AliasChoices("workout_id", "plan_id"))
    plan_id: Optional[uuid.UUID] = Field(None, validation_alias=AliasChoices("plan_id", "workout_id"))

    model_config = ConfigDict(from_attributes=True, populate_by_name=True, extra="ignore")


class TrainingLocation(str, Enum):
    HOME = "Дом"
    GYM = "Зал"


class WorkoutPlanBase(BaseModel):
    target_split: Optional[str] = None
    location: Optional[str] = "Дом"
    split_day: Optional[str] = None
    ai_rationale: Optional[str] = None
    status: Optional[str] = "planned"


class WorkoutPlanCreate(WorkoutPlanBase):
    pass


class WorkoutPlanResponse(WorkoutPlanBase):
    id: uuid.UUID
    user_id: Optional[uuid.UUID] = None
    created_at: Optional[datetime] = None
    exercises: List[WorkoutExerciseResponse] = Field(default_factory=list)

    model_config = ConfigDict(from_attributes=True, populate_by_name=True, extra="ignore")


class WorkoutGeneratePayload(BaseModel):
    location: TrainingLocation = TrainingLocation.HOME
    split_day: str = "День 1: Функционал"
    target_split: Optional[str] = None

class ExerciseAlternativeReason(str, Enum):
    JOINT_PAIN = "joint_pain"          # Боль / дискомфорт в суставах или связках
    NO_AXIAL_LOAD = "no_axial_load"    # Исключить осевую компрессию позвоночника
    EQUIPMENT_BUSY = "equipment_busy"  # Занят тренажер / нет снаряда
    TOO_INTENSE = "too_intense"        # Слишком тяжело / утомление

class AlternativeRequest(BaseModel):
    reason: ExerciseAlternativeReason = ExerciseAlternativeReason.JOINT_PAIN
    custom_note: Optional[str] = None

class AlternativeResponse(BaseModel):
    original_exercise: str
    alternative_exercise: str
    reason: Optional[str] = None
    target_muscle_groups: Optional[List[str]] = None
    rpe_target: Optional[float] = None

class ExerciseAlternativeItem(BaseModel):
    exercise_name: str
    exercise_type: str = "hypertrophy"
    sets: int = 3
    reps_or_duration: str = "10-12"
    rpe_target: int = 7
    target_muscle_groups: List[str] = Field(default_factory=list)
    biomechanical_rationale: str

class ExerciseSwapPayload(BaseModel):
    exercise_name: str
    exercise_type: str = "hypertrophy"
    sets: int = 3
    reps_or_duration: str = "10-12"
    rpe_target: int = 7
    target_muscle_groups: List[str] = Field(default_factory=list)

class TrainingResearchBase(BaseModel):
    title: str
    source_url: Optional[str] = None
    key_takeaways: str
    tags: List[str] = Field(default_factory=list)

class TrainingResearchCreate(TrainingResearchBase):
    pass

class TrainingResearchResponse(TrainingResearchBase):
    id: uuid.UUID
    pmid: Optional[str] = None
    abstract: Optional[str] = None
    applied_to_protocol: bool = False

    model_config = ConfigDict(from_attributes=True)


class CoachActionItem(BaseModel):
    type: str
    exercise_id: Optional[uuid.UUID] = None
    name: Optional[str] = None
    sets: Optional[int] = None
    reps: Optional[str] = None
    rpe: Optional[int] = None
    targets: List[str] = Field(default_factory=list)

class CoachActionResponse(BaseModel):
    coach_response: str
    actions: List[CoachActionItem] = Field(default_factory=list)

class NutritionIngredient(BaseModel):
    name: str
    weight_g: float
    protein: float
    fat: float
    carbs: float

class NutritionScanResponse(BaseModel):
    name: str
    weight_g: float
    calories: float
    protein: float
    fat: float
    carbs: float
    confidence: float
    ingredients: List[NutritionIngredient] = Field(default_factory=list)

class KineticsChatLogResponse(BaseModel):
    id: uuid.UUID
    module: str
    role: str
    message: str
    log_date: date

    model_config = ConfigDict(from_attributes=True)

class KineticsInsightResponse(BaseModel):
    id: uuid.UUID
    module: str
    period_start: date
    period_end: date
    insight_text: str

    model_config = ConfigDict(from_attributes=True)


class MesocycleGeneratePayload(BaseModel):
    start_date: date
    days_of_week: List[int] = Field(..., description="0=ПН, 1=ВТ, ..., 6=ВС")
    target_split: str = "Push/Pull/Legs"
    location: str = "Дом"
    include_deload: bool = True
    weeks_count: int = 4

class MesocycleExercise(BaseModel):
    name: str
    target_muscle_groups: List[str]
    sets: int
    reps_or_duration: str
    rpe_target: int
    type: str = "hypertrophy"

class MesocycleSession(BaseModel):
    target_split: str
    ai_rationale: str
    exercises: List[MesocycleExercise]

class MesocyclePlan(BaseModel):
    sessions: List[MesocycleSession]


class NutritionMealCreate(BaseModel):
    meal_date: date
    time_str: str = "12:00"
    name: str
    weight_g: float
    calories: float
    protein: float
    fat: float
    carbs: float
    ingredients: List[NutritionIngredient] = Field(default_factory=list)

class NutritionMealResponse(NutritionMealCreate):
    id: uuid.UUID
    user_id: uuid.UUID
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)

class DailyNutritionSummaryResponse(BaseModel):
    date: date
    total_calories: float
    total_protein: float
    total_fat: float
    total_carbs: float
    meals: List[NutritionMealResponse]

from enum import Enum
class SetType(str, Enum):
    WARMUP = "W"
    NORMAL = "N"
    DROPSET = "D"
    FAILURE = "F"

class WorkoutSetCreate(BaseModel):
    set_number: int
    set_type: SetType = SetType.NORMAL
    weight_kg: float = 0.0
    reps: int = 0
    rpe: Optional[float] = None
    is_completed: bool = False
    previous_weight_kg: Optional[float] = None
    previous_reps: Optional[int] = None

class WorkoutSetUpdate(BaseModel):
    weight_kg: Optional[float] = None
    reps: Optional[int] = None
    rpe: Optional[float] = None
    is_completed: Optional[bool] = None

class WorkoutSetResponse(WorkoutSetCreate):
    id: uuid.UUID
    exercise_id: uuid.UUID

    model_config = ConfigDict(from_attributes=True)


class WorkoutDuplicateRequest(BaseModel):
    target_date: str
    apply_overload: bool = False
    overload_increment_kg: float = 1.25



class OneRMDataPoint(BaseModel):
    date: str
    one_rm_kg: float

class ExerciseOneRM(BaseModel):
    exercise_name: str
    history: List[OneRMDataPoint]

class TonnageDataPoint(BaseModel):
    week_start: str
    muscle_group: str
    tonnage_kg: float

class OverloadAnalyticsResponse(BaseModel):
    one_rm_top_exercises: List[ExerciseOneRM]
    weekly_tonnage: List[TonnageDataPoint]



class BiometricsCreate(BaseModel):
    weight_kg: Optional[float] = None
    sleep_hours: Optional[float] = None
    fatigue_score: Optional[int] = None
    notes: Optional[str] = None

class RecoveryDataPoint(BaseModel):
    date: str
    tonnage_kg: float
    calories: int
    protein_g: int
    sleep_hours: float
    fatigue_score: int
    recovery_score: float

class RecoveryAnalyticsResponse(BaseModel):
    timeline: List[RecoveryDataPoint]
    average_sleep: float
    average_calories: float
    total_tonnage: float
    current_recovery_score: float

